pkg/controller/job/job_controller_test.go

```diff
@@ -1269,3 +1258,78 @@ func bumpResourceVersion(obj metav1.Object) {
 	ver, _ := strconv.ParseInt(obj.GetResourceVersion(), 10, 32)
 	obj.SetResourceVersion(strconv.FormatInt(ver+1, 10))
 }
+
+type pods struct {
+	pending int32
+	active  int32
+	succeed int32
+	failed  int32
+}
+
+func TestJobBackoffReset(t *testing.T) {
+	testCases := map[string]struct {
+		// job setup
+		parallelism  int32
+		completions  int32
+		backoffLimit int32
+
+		// pod setup - each row is additive!
+		pods []pods
+	}{
+		"parallelism=1": {
+			1, 2, 1,
+			[]pods{
+				{0, 1, 0, 1},
+				{0, 0, 1, 0},
+			},
+		},
+		"parallelism=2 (just failure)": {
+			2, 2, 1,
+			[]pods{
+				{0, 2, 0, 1},
+				{0, 0, 1, 0},
+			},
+		},
```

Comment:
@foxish if I add the 3rd test here, with +1 failure and +1 success this is where it won't get reseted, and I'm not 100% convinced it should. /cc @clamoriniere1A

Reply:
My reasoning is that with small parallelism it should not be a problem, but what about big ones, esp. bigger than backoff limit? Or maybe we should ensure the backoff limit can't be lower than the parallelism, at least.

Reply:
That makes sense, maybe we should set the `backofflimit` to `min(6, parallelism)`?

Reply:
On second thought - maybe that's independent. With both failure and success at the same time, maybe we should leave the backoff time unchanged and retry?

Reply:
I've created https://github.com/kubernetes/kubernetes/issues/56853 to discuss this and have it solved for 1.10. Here I've added a separate commit to change the completions to 3 in that e2e to get rid of that flake.

Reply:
I want to leave that unit test for the backoff limit resetting, in place.